lib/class-wp-theme-json-gutenberg.php

```diff
@@ -2632,10 +2626,14 @@ private static function get_block_nodes( $theme_json, $selectors = array() ) {
 			$variation_selectors = array();
 			if ( isset( $node['variations'] ) ) {
 				foreach ( $node['variations'] as $variation => $node ) {
-					$variation_selectors[] = array(
-						'path'     => array( 'styles', 'blocks', $name, 'variations', $variation ),
-						'selector' => $selectors[ $name ]['styleVariations'][ $variation ],
-					);
+					// Only add registered variations to this node's variations to
+					// avoid CSS generated without a selector.
+					if ( isset( $selectors[ $name ]['styleVariations'][ $variation ] ) ) {
```

Comment:
So, what this PR does is:

- do not remove the **unregistered** variations when we build the theme.json in memory
- but make sure we only process the **registered** ones when we use them

This is the kind of thing that I mentioned: we are not removing sanitization but moving it **everywhere**. Each time we have to work with them, we need to make sure to use the ones that are registered.